### PR TITLE
[FLINK-23420][table-runtime] LinkedListSerializer now checks for null elements in the list

### PR DESCRIPTION
**Pull request by tsreaper**
## What is the purpose of the change

`LinkedListSerializer` in the table runtime module currently is not aware of the null elements in the list. This will cause NPE in some functions like the `LAG` aggregation function.

This PR fixes this issue and at the mean time keep the compatibility of the states.

## Brief change log

 - LinkedListSerializer now checks for null elements in the list

## Verifying this change

This change added tests and can be verified by running the added tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: **YES**
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by tsreaper** [comment]
@flinkbot run azure

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java [hunk 1]
```diff
@@ -186,12 +267,28 @@ public int getCurrentOuterSnapshotVersion() {
...
+        protected void readOuterSnapshot(
+                int readOuterSnapshotVersion, DataInputView in, ClassLoader userCodeClassLoader) {
+            readVersion = readOuterSnapshotVersion;
+        }
+
+        @Override
+        protected OuterSchemaCompatibility resolveOuterSchemaCompatibility(
+                LinkedListSerializer<T> newSerializer) {
+            if (readVersion <= LAST_NULL_NOT_CHECKED_VERSION) {
+                return OuterSchemaCompatibility.COMPATIBLE_AFTER_MIGRATION;
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java:280]: We could consider `NullableSerializer.checkIfNullSupported` too.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java [hunk 2]
```diff
@@ -186,12 +267,28 @@ public int getCurrentOuterSnapshotVersion() {
...
+        @Override
+        protected void readOuterSnapshot(
+                int readOuterSnapshotVersion, DataInputView in, ClassLoader userCodeClassLoader) {
+            readVersion = readOuterSnapshotVersion;
+        }
+
+        @Override
+        protected OuterSchemaCompatibility resolveOuterSchemaCompatibility(
+                LinkedListSerializer<T> newSerializer) {
+            if (readVersion <= LAST_NULL_NOT_CHECKED_VERSION) {
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java:279]:
```
private boolean isCompatible() {
   return readVersion > LAST_NULL_NOT_CHECKED_VERSION || NullableSerializer.checkIfNullSupported(eleSer);
}
```

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java [hunk 3]
```diff
@@ -39,18 +39,27 @@
...
 
+    // legacy, don't touch until we drop support for 1.9 savepoints
     private static final long serialVersionUID = 1L;
 
-    /** The serializer for the elements of the list. */
+    // The serializer for the elements of the list.
     private final TypeSerializer<T> elementSerializer;
 
+    // When legacy mode is enabled serializer will not check for null values.
+    private final boolean legacyModeEnabled;
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java:49]: `nullSafe`?

**Review by JingsongLi — CHANGES_REQUESTED** [review]
Thanks! Left some comments.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java (hunk 1 above)
Comment (JingsongLi) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java:280]: Or you can use `MaskUtils` and ignore `checkIfNullSupported`. This way will not waste much space.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java (hunk 3 above)
Comment (JingsongLi) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java:49]: Use `MaskUtils` and ignore this comment.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java (hunk 2 above)
Comment (JingsongLi) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java:279]: It is better to extract a method.

**Comment by tsreaper** [comment]
After discussing with @JingsongLi , the biggest problem of the current implementation is the overhead of the null boolean before each element. We're going to use `MaskUtils#writeMask` like `RowSerializer` instead.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java [hunk 4]
```diff
@@ -110,17 +124,40 @@ public int getLength() {
...
         for (T element : list) {
-            elementSerializer.serialize(element, target);
+            if (element != null) {
+                elementSerializer.serialize(element, target);
+            }
         }
     }
 
+    private boolean[] getNullMask(LinkedList<T> list) {
+        boolean[] mask = new boolean[list.size()];
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java:138]: I think we should reuse this mask.

Comment (tsreaper) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java:138]: I suppose by "reusing" you're meant to not creating a boolean array every time and change its size when the length of list grows.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java [hunk 5]
```diff
@@ -110,17 +122,42 @@ public int getLength() {
...
         for (T element : list) {
-            elementSerializer.serialize(element, target);
+            if (element != null) {
+                elementSerializer.serialize(element, target);
+            }
         }
     }
 
+    private boolean[] getNullMask(LinkedList<T> list) {
+        if (reuseMask == null || reuseMask.length != list.size()) {
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/typeutils/LinkedListSerializer.java:136]: `reuseMask.length < list.size()`
We can introduce a `void writeMask(boolean[] mask, int length, DataOutputView target)` to `MaskUtils`.

**Comment by tsreaper** [comment]
@flinkbot run azure

**Review by JingsongLi — APPROVED** [review]
Thanks for the update! Looks good to me.

**Comment by JingsongLi** [comment]
@tsreaper Can you create JIRA for 1.13?

**Pull request merged**